### PR TITLE
implement search

### PR DESCRIPTION
**Pull request by tomerd**
motivation: continue to implement package collections business logic

changes:
* implement in memory search for packages and targets
* performance work on "list" by making decoding in parallel (TBD)
* enable performance tests

### CODE REVIEW

File: Sources/PackageCollections/Storage/PackageCollectionsStorage.swift [hunk 1]
```diff
@@ -226,25 +225,87 @@ final class SQLitePackageCollectionsStorage: PackageCollectionsStorage, Closable
                     }
                 }
 
-                // TODO: consider some diagnostics / warning for invalid data
-                let collections = blobs.compactMap { data -> PackageCollectionsModel.Collection? in
-                    try? self.jsonDecoder.decode(PackageCollectionsModel.Collection.self, from: data)
+                // FIXME: is there a better way?
```
Comment (tomerd) [inline Sources/PackageCollections/Storage/PackageCollectionsStorage.swift:228]: 👀  this is an interesting one. I put in a workaround of parallelizing the decoding for large enough set, but the out of-the-box performance is JSONDecoder is not impressive

**Comment by tomerd** [comment]
@yim-lee @abertelrud @neonichu

**Comment by tomerd** [comment]
@swift-ci please test

**Review by neonichu — APPROVED** [review]

File: Sources/PackageCollections/Storage/PackageCollectionsStorage.swift (hunk 1 above)
Comment (yim-lee) [reply Sources/PackageCollections/Storage/PackageCollectionsStorage.swift:228]: 👍

File: Sources/PackageCollections/Storage/PackageCollectionsStorage.swift [hunk 2]
```diff
@@ -269,12 +330,79 @@ final class SQLitePackageCollectionsStorage: PackageCollectionsStorage, Closable
...
+        let query = query.lowercased()
+
+        self.list(identifiers: identifiers) { result in
+            switch result {
+            case .failure(let error):
+                callback(.failure(error))
+            case .success(let groups):
+                let collectionsPackages = groups.reduce([PackageCollectionsModel.CollectionIdentifier: [(target: PackageCollectionsModel.PackageTarget, package: PackageCollectionsModel.Collection.Package)]]()) { partial, group in
+                    var map = partial
+                    // var result = [(PackageGroup.Package.Target, PackageGroup.Package)]()
```
Comment (yim-lee) [inline Sources/PackageCollections/Storage/PackageCollectionsStorage.swift:347]: nit: delete?

**Review by yim-lee — APPROVED** [review]
Looks great.

We should revisit in the future to make sure the search APIs are returning relevant info. The logic that puts the results together is too complex (not to your fault). If some of the info is not used then we should remove it and simplify the logic if we can.

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
merging this so that CLI changes can progress. follow up PR on search performance incoming

**Pull request merged**